test/e2e/network/service.go

```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
 		framework.ExpectNoError(verifyServeHostnameServiceUp(cs, ns, host, podNames2, svc2IP, servicePort))
 	})
 
+	ginkgo.It("should be able to preserve UDP traffic when server pod cycles [Slow]", func() {
+		serviceName := "clusterip-test"
+		serverPod1Name := "server-1"
+		serverPod2Name := "server-2"
+
+		ns := f.Namespace.Name
+
+		nodeIP, err := e2enode.PickIP(cs) // for later
+		framework.ExpectNoError(err)
+
+		udpJig := e2eservice.NewTestJig(cs, ns, serviceName)
+		//udpService, err := udpJig.CreateUDPService(nil)
+		udpService, err := udpJig.CreateUDPService(func(svc *v1.Service) {
+			svc.Spec.Type = v1.ServiceTypeNodePort
+			svc.Spec.Ports = []v1.ServicePort{
+				{Port: 80, Name: "http", Protocol: v1.ProtocolUDP, TargetPort: intstr.FromInt(80)},
+			}
+		})
+		framework.ExpectNoError(err)
+
+		serverPod1 := newAgnhostPod(serverPod1Name, "netexec", fmt.Sprintf("--udp-port=%d", 80))
+		serverPod1.Labels = udpJig.Labels
+		_, err = cs.CoreV1().Pods(ns).Create(context.TODO(), serverPod1, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, serverPod1.Name, f.Namespace.Name, framework.PodStartTimeout))
+
+		errorChannel := make(chan error)
+		go continuousEcho(nodeIP, int(udpService.Spec.Ports[0].NodePort), 3*time.Second, 20, errorChannel)
```

Comment:
can we avoid duplicating the pokeUDP() into a new function continuousEcho() and use it directly in a for loop?

```go
go func() {
     defer GinkgoRecover()

  	for i := 0; i < maxAttempts; i++ {
             pokeUDP(host, port, "echo hello", &UDPPokeParams{
			Timeout:  3 * time.Second,
			Response: "hello",
		})
		if result.Status != UDPSuccess {
			numErrors++
		}

       }
  	err = nil
	if numErrors >= 6 {
		err = fmt.Errorf("Too many Errors in continuous echo")
	}
	errorChannel <- err
}
```

Reply:
no, this wont exercise the behaviour every call to pokeUDP will generate a difference source port and create a different conntrack entry. I needed to modify pokeUDP so that continuousEcho() is the same stream of packets with the same source port, the service needs to correctly change the flow of traffic to the new pod

Reply:
ahh good point, thank you, totally missed that